### PR TITLE
Server-side builds (version 0.1.0)

### LINKED ISSUE #413: Generate test database vs. HTML files
To aid in tests, we should do the following (as suggested in https://github.com/foolip/mdn-bcd-collector/issues/376#issuecomment-685021436):

- Generate a database (JSON file) of tests per BCD data path.
- Let the server use that database to generate the test pages, keeping the output of those pages as trivial as possible. The browser should not need to query and interpret MANIFEST.json to run tests.

### LINKED ISSUE #414: Move backend endpoints?
The `/api/` URL contains two separate pieces of functionality in it -- the API tests, and the API endpoints.  We should rename the API endpoint URLs to something else (for example, `/backend/` or just leave them as-is).  Either that, or move the tests themselves to a new URL, which we could do considering #413.

### LINKED ISSUE #417: Stylize pages
I was thinking that we should give the homepage, reports page, and individual test completion pages a little bit of styling to make them look more attractive.  What do you think, @foolip?

### LINKED ISSUE #418: Console API doesn't post updates correctly
In BCD, the Console API is referred to as `api.Console` -- however, the WebIDL actually defines `console` as a namespace with no capitalization.  Since the identifiers are case-sensitive, this causes no updated results.

There's a few solutions I can think of:
- Rename the identifier in the build script -- hacky and special-casing, but it'll do the trick at least
- Rename the identifier in update-bcd -- also hacky and special-casing
- Make identifiers case-insensitive -- could cause unforseen issues in BCD

### LINKED ISSUE #420: Generate endpoints on the fly in backend
> Do we need manifest.json separately from test.json, shouldn't the backend be able to derive everything it needs from just the list of tests?

_Originally posted by @foolip in https://github.com/foolip/mdn-bcd-collector/pull/416#discussion_r484376238_

Additional comments:
https://github.com/foolip/mdn-bcd-collector/pull/416#discussion_r484440344
https://github.com/foolip/mdn-bcd-collector/pull/416#discussion_r484450944

### LINKED ISSUE #437: Use <datalist> for individual test pages
As a follow-up to https://github.com/foolip/mdn-bcd-collector/issues/376, it would be nice to use an `<input list>` with `<datalist>` instead of a `<select>`, as this makes it easier to find the right endpoint by typing part of an API name.

### PR DESCRIPTION
**Pull request by vinyldarkscratch**
This PR provides a huge changeset that qualifies for the bump of a minor version.  Primarily, this PR re-introduces server-side builds, and performs a number of optimizations.

Updates include the following:

- Web UI
  - Add a little content to the homepage
  - Add CSS styling to the pages
  - The dropdown is now a `<datalist>`, allowing for type-to-search
  - Individual tests for APIs can be tested under a specified scope
  - There is now a dedicated 404 page
- Build/Generation
  - All tests are compiled server-side once again
  - All HTML is now generated using EJS templates
  - `MANIFESTS.json` is now reduced to just the tests, and is moved to `tests.json`
  - Endpoints are generated by the server
- Test Runtime
  - All the tests are now under `/tests/` to separate it from the backend API endpoints
  - Tests are both displayed and reported to the server in one go, with a link to submit a PR to the results GitHub repo
  - Promises are no longer used for the worker and service worker; thus, core-js is no longer a dependency
- Tests
  - `api.console` is now mapped to `api.Console`
  - The `toString` operation is now added
  - Added support custom API tests for non-members (i.e. `api.AbortSignal.abort_event` or `api.BaseAudioContext.createPeriodicWave.disableNormalisation_supported`)
  - Added test for `api.DOMTokenList.remove_whitespace_and_duplicates`
- Misc.
  - Nodemon is now used to auto-reload the server when any files change

Fixes #413, fixes #414, fixes #417, fixes #418, fixes #420, fixes #437.

### CODE REVIEW

**Comment by vinyldarkscratch** [comment]
I'm requesting your review, @foolip, since this is a big change, and I want to make sure it's all good before we merge and deploy!

File: app.js
Comment (foolip) [inline app.js:76]: Let's drop this. A big comment at the top is annoying when viewing source to understand the test. If I hear that we need this even for generated output then we can add it back at the bottom.

**Review by foolip — COMMENTED** [review]
This isn't my full review, I'll take a look at this next week. Very exciting change though, I think this will make things much easier to under for BCD reviewers.

File: app.js
Comment (vinyldarkscratch) [reply app.js:76]: Okay, I'll go ahead and drop it!  I think it'll be all fine, since there's copyright notices elsewhere.

**Comment by vinyldarkscratch** [comment]
Whew!  I have now added prefix support into server-side builds, so we can get prefix support once again!  It functions in a similar manner to how the prefix testing worked before -- test each piece with each prefix until support is found, then move onto the next -- with different prefixed variations built into the manifest.  Unfortunately, this makes the manifest file very long and very large as a result -- I've separated it into two files.

Long story short, I think this is now ready for deploying!

File: static/resources/harness.js
Comment (foolip) [inline static/resources/harness.js:344]: In which browser does `document.head.appendChild(css)` throw an exception? Any exception shouldn't be due to the CSS failing to load, which happens asynchronously, but more likely because `document.head` doesn't exist. I'm guessing this happened on IE<9, based on https://developer.mozilla.org/en-US/docs/Web/API/Document/head?

File: build.js
Comment (foolip) [inline build.js:664]: Putting this after the `await fs.readFile` would be good. Or, if this will all go away by not using core-js, a TODO would be fine.

File: build.js
Comment (foolip) [inline build.js:670]: Not reviewing this in detail, but I note that it's not a very long function, so the server should be able to compute all of this information from the tests listing at runtime. (See also comment in app.js about `/tests/*` for how I think this can change.)

File: build.js
Comment (foolip) [inline build.js:734]:
```suggestion
  const tests = Object.assign({}, IDLTests, CSSTests);
```

To avoid modifying `IDLTests` in a way that's not obvious.

File: build.js
Comment (foolip) [inline build.js:732]: By passing in `bcd` here, we should have the information we need about which prefixes to test for.

File: non-standard.idl
Comment (foolip) [inline non-standard.idl:32]: Can you split out all updates to non-standard.idl into its own PR?

File: find-missing.js
Comment (foolip) [inline find-missing.js:26]: Good stuff! From that list one can see a lot that can be added to non-standard.idl. But it's best to do so separately from this overhaul.

File: .eslintrc.json
Comment (foolip) [inline .eslintrc.json:3]: By now it may be incredibly hard to untangle, but doing the reformatting resulting from this and other linter changes separately from the backend overhaul would be great.

What I'd do: copy over the relevant changes to package.json and .eslintrc.json changes to a new branch, and reformat using `--fix` and cherry-picking over any changes that can't be made automatically.

When that changes is on main, merge in the main branch and resolve any conflicts in favor of what's on this branch. What remains should be the real changes on this branch.

**Review by foolip — COMMENTED** [review]
Submitting a partial review, I stopped half way through build.js.

File: static/resources/harness.js
Comment (vinyldarkscratch) [reply static/resources/harness.js:344]: It was actually early versions of Chrome (like Chrome 4), which led me to believe that it may occur in other browsers as well (like IE).  I tried with `document.getElementsByTagName('head')[0]` as well, but had no luck there.

File: .eslintrc.json
Comment (vinyldarkscratch) [reply .eslintrc.json:11]: I thought I had to redefine the entire rule if I was making any changes to it, but it looks like rule options will combine!

File: DESIGN.md
Comment (vinyldarkscratch) [reply DESIGN.md:58]: I wanted to put some sort of filler test down in the document regardless of whether it actually functioned, just to show how the additional tests function, and then come back to replace it with a working one later.

File: app.js
Comment (vinyldarkscratch) [reply app.js:49]: Yeah, the list of tests should give the backend everything that it needs!  I'm thinking to save this for another time.

File: .eslintrc.json
Comment (vinyldarkscratch) [reply .eslintrc.json:3]: Cherry-picked as #421!

File: unittest/unit/tests.js
Comment (vinyldarkscratch) [reply unittest/unit/tests.js:129]: Collapsing the tests does sound nice, it ends up making the amount of tests in a single file a little too much, causing browser freezes (especially on older browsers).  I'm also not sure how to properly hook it up with our report function to submit PRs for all of the data to the results repo.  Will have to come back to this one at a later time.

Comment (foolip) [reply unittest/unit/tests.js]: Moved to https://github.com/foolip/mdn-bcd-collector/issues/430.

File: build.js
Comment (foolip) [reply build.js:712]: Moved to https://github.com/foolip/mdn-bcd-collector/issues/431.

File: .eslintrc.json
Comment (foolip) [reply .eslintrc.json]: Thanks!

Comment (foolip) [reply .eslintrc.json]: Done in https://github.com/foolip/mdn-bcd-collector/pull/421, thanks!

File: find-missing.js
Comment (foolip) [reply find-missing.js]: Done in https://github.com/foolip/mdn-bcd-collector/pull/422.

File: non-standard.idl
Comment (foolip) [reply non-standard.idl]: Done in https://github.com/foolip/mdn-bcd-collector/pull/427.

File: build.js
Comment (foolip) [reply build.js:712]: Moved to https://github.com/foolip/mdn-bcd-collector/issues/431.

File: app.js
Comment (foolip) [reply app.js]: Moved to https://github.com/foolip/mdn-bcd-collector/issues/420.

File: build.js
Comment (foolip) [reply build.js:166]: Moved to https://github.com/foolip/mdn-bcd-collector/issues/431.

File: DESIGN.md
Comment (foolip) [reply DESIGN.md:58]: Alright. I think `api.DOMTokenList.remove_whitespace_and_duplicates` is a good candidate for a simple custom test for illustration purposes:
```js
var elm = document.createElement('b');
elm.className = ' foo bar foo ';
elm.classList.remove('bar');
return elm.className === 'foo';
```

File: static/resources/harness.js
Comment (foolip) [reply static/resources/harness.js]: I see. Catching and ignoring the error is fine. Only the comment needs updating, something like "If inserting the link element fails" would be more accurate about where the exception is coming from. (The consequence will be the CSS won't load, though.)

File: DESIGN.md
Comment (vinyldarkscratch) [reply DESIGN.md:58]: Sure thing, added!

File: build.js
Comment (foolip) [inline build.js:84]: I've filed https://github.com/foolip/mdn-bcd-collector/issues/432 to undo weird Prettier formatting like this.

File: static/resources/harness.js
Comment (foolip) [inline static/resources/harness.js:152]: This isn't new, but I'm nervous about doing really anything at all while the tests are running. Let's come back to this when moving more of the code generation to the server side.

File: static/resources/harness.js
Comment (foolip) [inline static/resources/harness.js:172]: Not in this PR, but the setup for worker tests I'm imaging when most of the of work is done on the server side is that a single large script is run in a worker, and then the results are all posted back in one go. In the window context there wouldn't be any need to keep track of pending tests.

File: static/resources/harness.js
Comment (foolip) [inline static/resources/harness.js]:
```suggestion
        // If document.head isn't supported, no CSS will be loaded. That's OK.
```

File: tests.js
Comment (foolip) [inline tests.js:34]: In a follow-up, maybe `/api/?scope=Window`, modulo naming of "scope" discussed elsewhere. Not only interfaces are being tested any longer, namespaces and attributes/functions on the global scope are too.

File: tests.js
Comment (foolip) [inline tests.js:36]: Can both 'Worker' and 'DedicatedWorker' happen? It seems like it should be one or the other.

File: tests.js
Comment (foolip) [inline tests.js:57]: This won't work. `x in y` where `y` is an array will only be true if `x` is an array index (string or number) or if it's some Array builtin like `length` or `push`. Not the cases you want to check for, I presume.

File: tests.js
Comment (foolip) [inline tests.js:69]: What does the empty string mean here? I think it's a marker for "run this test by default", is that right? Being more explicit about this would be good.

File: tests.js
Comment (foolip) [inline tests.js:92]: What if `index` is -1 (before we add 1)? Isn't that an error case that we should do something about? As it is, will we end up running the same test in an endless cycle?

File: tests.js
Comment (foolip) [inline tests.js:105]: I've filed https://github.com/foolip/mdn-bcd-collector/issues/435 to perhaps simplify this later. What I did was a bit too clever IMHO, and continuing to maintain the complexity like here is unfortunate.

File: tests.js
Comment (foolip) [inline tests.js:115]:
```suggestion
    let idents;
```

It's assigned in both branches, let's be explicit that this empty array is never used.

File: tests.js
Comment (foolip) [inline tests.js:171]: This is a place where I expect resolving https://github.com/foolip/mdn-bcd-collector/issues/376 will change things a lot. Rather than feeding input to the client-side test generator here, we should just output stuff like:
```js
bcd.addTest('api.Document', function() {
  return 'Document' in window;
});
```

But not in this PR :)

**Review by foolip — COMMENTED** [review]
Finally got through the whole change. Lots of comments, please enjoy :)

Let's punt as much as possible to issues where things aren't regressed by this change.

File: static/resources/harness.js
Comment (vinyldarkscratch) [reply static/resources/harness.js:110]: The reason for `eval` is to be able to send the functions to the workers -- all messages sent to workers must be serializable, so passing in functions wouldn't work, unfortunately.

File: static/resources/harness.js
Comment (vinyldarkscratch) [reply static/resources/harness.js:111]: This is a part of the addition of `bcd.testConstructor`, which returns an object (that can include a message).

File: tests.js
Comment (vinyldarkscratch) [reply tests.js:36]: Some interfaces mark exposure for `Worker`, while others for `DedicatedWorker`.  From what I've gathered, they're both the same however.

File: package.json
Comment (vinyldarkscratch) [reply package.json:41]: Overall it shouldn't interfere with GAE, but just in case, I'll split this into `start-dev` instead.  Nodemon watches for any file changes and automatically reloads the server accordingly -- SUPER helpful for development purposes.

File: static/index.html
Comment (vinyldarkscratch) [reply static/index.html]: Since it's within the scope of server-side builds, I decided to actually tackle this, and hook up EJS to build the pages.  All pages -- including the homepage, results, and the tests themselves -- are now built with the EJS templating system!

File: static/resources/harness.js
Comment (vinyldarkscratch) [reply static/resources/harness.js:152]: The `updateStatus()` isn't really working due to how much is going on in the JavaScript already (the DOM isn't updating), so these can be taken out safely.

File: tests.js
Comment (vinyldarkscratch) [reply tests.js:34]: I just pushed f8934be379e465cdc78a2d514d0f3a6d6c164380 which should provide some assistance for this.  In that commit, I added `?reportToServer=true`, which will tell the website to send the data to the server rather than display it.  It no longer counts on a test being a "main endpoint" to determine if it should report, which works towards eliminating main endpoints altogether.

File: views/index.ejs
Comment (foolip) [inline views/index.ejs:71]: Could this use the protocol of the current page instead? For localhost 'https:' will never make sense, so showing the option then isn't really useful.

File: views/layout.ejs
Comment (foolip) [inline views/layout.ejs:4]: Let's omit copyright notices from the generated pages. I shouldn't have included it in the first place :)

File: views/tests.ejs
Comment (foolip) [inline views/tests.ejs:28]: Can you add a TODO linking to https://github.com/foolip/mdn-bcd-collector/issues/407 here?

**Review by foolip — COMMENTED** [review]
All done reviewing, phew!

I think this looks pretty good now, I'd like to get it landed so we can keep iterating in smaller PRs. I've tried deploying https://staging-dot-mdn-bcd-collector.appspot.com/ but something is broken.

Once we have a working staging deployment, it would be good to compare the collected results before/after this change, to see if there are unintended changes in there.

**Comment by foolip** [comment]
It looks like the 500 error on https://staging-dot-mdn-bcd-collector.appspot.com/ is because the 256 MB limit is being exceeded. @vinyldarkscratch does that give any clues? Feel free to push changes to the staging branch to narrow down what it might be that causes this.

File: nodemon.json
Comment (vinyldarkscratch) [reply nodemon.json:5]: It's because the server itself doesn't need to reload for that -- just the browser.

File: tests.js
Comment (vinyldarkscratch) [reply tests.js:93]: Fixed down below!  If the `index` is 0 (`-1 + 1`), we return `null` now, to prevent an endless loop.

File: build.js
Comment (vinyldarkscratch) [reply build.js]: Yeah, `exposure` was certainly misused here.  I'm now adding a `category` property, which will also help future-proof for `javascript` feature testing or other categories.

File: static/resources/harness.js
Comment (vinyldarkscratch) [reply static/resources/harness.js]: I've now fixed up `exposure` and removed this `runCSS()` function entirely!

File: build.js
Comment (vinyldarkscratch) [reply build.js:680]: `CSS.supports` is kind of a special case in this sense, as it's the only instance of "how do we test it".  We could, perhaps, call it "method" in this case?

**Comment by vinyldarkscratch** [comment]
> It looks like the 500 error on https://staging-dot-mdn-bcd-collector.appspot.com/ is because the 256 MB limit is being exceeded. @vinyldarkscratch does that give any clues? Feel free to push changes to the staging branch to narrow down what it might be that causes this.

What is the 256 MB limit for?  Is it storage space, or how large the response from the server is?  (Unfortunately my knowledge of GAE is limited...)

**Comment by vinyldarkscratch** [comment]
Great news: I was able to eliminate "main endpoints" in an effective manner, so we now only have endpoints for the individual tests!  My solution was to have the script both submit AND display the results, no redirects, with a link at the top to submit the results to the GitHub repo.

**Comment by foolip** [comment]
@vinyldarkscratch other than this not working on AppEngine, is this ready for final review, or do you have more changes planned?

**Comment by vinyldarkscratch** [comment]
I'd say this is ready for final review!

**Comment by vinyldarkscratch** [comment]
I want to try to integrate the datalist if possible, and simplify the code some if I can.  I should be ready for a final review soon, though!

**Comment by vinyldarkscratch** [comment]
Okay, this is now ready for final review, @foolip!

File: build.js
Comment (foolip) [reply build.js:621]: That would be fine, yeah. I won't block this PR on fixing this though, I've linked it in https://github.com/foolip/mdn-bcd-collector/issues/375 to not forget.

File: unittest/app/app.js
Comment (foolip) [inline unittest/app/app.js:139]: Good testing!

**Review by foolip — APPROVED** [review]
Let's get this merged!!!

**Pull request merged**
